### ISSUE TITLE
spec: use TXT dns records for discovery

### ISSUE THREAD

**Issue by philips** [appc/spec#3]
_From @philips on December 8, 2014 21:3_

@blalor said "I feel like there should be a DNS component to this, as well. Perhaps a TXT record?" in #210.

Lets discuss that topic here.

/cc @geekgonecrazy

_Copied from original issue: coreos/rocket#242_

**Comment by dapithor** [comment]
Not trying to bump an old item but this is somewhat how I was thinking to do this with SRV (per my previous comments). I didn't articulate it well enough though... not sure how close you all work with mesos but for reference:
http://mesosphere.github.io/mesos-dns/docs/naming.html (which I see uses miekg's framework).

**Comment by philips** [comment]
@dapithor I don't understand how the mesos-dns tool helps here. Can you explain a bit further?

**Comment by geekgonecrazy** [comment]
@dapithor Maybe try explaining again?  

Because from what I see they are using the SRV records to point to an ip.  Which is what's supposed to be in an SRV record.

The problem with just an ip address is its still not telling rocket(or any tool in the future) where to grab the .aci file.  Its just telling it the subdomain that its on.  Since its fetching via http, it needs to be able to determine the full url.

**Comment by cdaylward** [comment]
I think there are two things going on here:

1. Should a DNS discovery mechanism be added to the spec?
2. Should the discovery _process_ be as prescriptive as it is? I really like the two discovery mechanisms as they are currently outlined, however I would like to see discovery reframed as "pluggable, ordered discovery strategies". Simple discovery is a strategy. Meta-discovery is a strategy. DNS discovery is a strategy. I believe it would be beneficial to outline their structure but not prescribe which ones must be used or in which order. This is a little OT here, should this be a separate discussion thread?

As far as using DNS goes, SRV is I believe half the story. Combining an [NAPTR record](http://en.wikipedia.org/wiki/NAPTR_record) with the use of SRV (or without) gives you resource location and URI rewrite rules. This could be conceivably used to turn label sets into resource locators.

**Comment by jonboulle** [comment]
2) is a distinct but good point that I would also like to address, mind filing another issue/PR for that?

**Comment by geekgonecrazy** [comment]
I agree, they should probably be pluggable.  But order of discovery methods would probably be an entirely separate discussion.

How common are NAPTR records?  I honestly can't say i've ever even seen that record type until now.  I checked both DNS providers i've used in the past, and neither of them even offer that record type.  

Looks intriguing though..

**Comment by cdaylward** [comment]
@jonboulle Not at all. Will do.

**Comment by cdaylward** [comment]
@geekgonecrazy NAPTR records are used extensively by VOIP carriers, their peers, and clients (I know this first-hand). I've also used them on other internal networks for other types of resource location (you can use it to select based on preferred protocol). The [RFC](https://www.ietf.org/rfc/rfc2915.txt) has examples of general use, e.g. (this one does not use the rewrite field)

```
 IN NAPTR 100  50  "s"  "z3950+I2L+I2C"     ""  _z3950._tcp.gatech.edu.
 IN NAPTR 100  50  "s"  "rcds+I2C"          ""  _rcds._udp.gatech.edu.
 IN NAPTR 100  50  "s"  "http+I2L+I2C+I2R"  ""  _http._tcp.gatech.edu.
```

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [spec] appc: use TXT dns record..." ]
>@geekgonecrazy NAPTR records are used extensively by VOIP carriers, their peers, and clients (I know this first-hand). I've also used them on other internal networks for other types of resource location (you can use it to select based on preferred protocol). The [RFC](https://www.ietf.org/rfc/rfc2915.txt) has examples of general use, e.g. (this one does not use the rewrite field)
>
>```
> IN NAPTR 100  50  "s"  "z3950+I2L+I2C"     ""  _z3950._tcp.gatech.edu.
> IN NAPTR 100  50  "s"  "rcds+I2C"          ""  _rcds._udp.gatech.edu.
> IN NAPTR 100  50  "s"  "http+I2L+I2C+I2R"  ""  _http._tcp.gatech.edu.
>```

Indeed NAPTR is just a normal record, except nobody uses it on the internet. I
don't think DNS server providers will allow you to add NAPTR record out of the
box.

As I said before, if you going to take the route of service discovery in the DNS
you probably best of designing your own record type. (To hell with those pesky
DNS service providers). And ... maybe al fallback to TXT (if you must).

/Miek

--
Miek Gieben

**Comment by dapithor** [comment]
Per @cdaylward mentioned that was my initial thoughts too.  Using a SRV record doesn't fulfill the request completely but it does provide purpose of listing a host, port, and protocol for a service of ACI's.  The second part I mentioned was to have a default location and if needed a corresponding TXT entry.  Of course I like the idea of a custom Resource Record per miekg's advice but to keep things "simple" and able to be used on most platforms SRV for service info and TXT for config if needed.  IMO having a custom RR you might as well just have a custom tool which is full circle to etcd like usage.  This allows an instance to run on a different port and maybe in future different protocol requirements and still leave the config alone (TXT record) unless no other port except 443 will be used or always doing an URI://location:port/{opt1}{opt2}{etc} is preferred then I say SRV is useless.

**Comment by miekg** [comment]
How is "lookup SRV" and *then* "lookup TXT" not custom? I'm not sure what the requirements here are? 

The only thing going for SRV and TXT lookups is that the resource records are well known. But doing combined SRV/TXT lookups to make stuff work also requires custom code, because no resolver will do this out of the box.

**Comment by dapithor** [comment]
Fair points but application logic needs to happen regardless if one record or N records.  So, if you look up anything logic needs to be in the app but SRV/TXT etc being built into most DNS servers/services alleviates the need to also have a custom dns server for a custom RR. 
Even with zeroconf/dns-sd (which is basically what I am advocating and should have clarified)  you still need a logic/library and tell it to use it? Am I too far off base ? 
My thoughts were basically DNS-SD which is SRV/TXT in this wiki:
http://en.wikipedia.org/wiki/Zero-configuration_networking#DNS-based_service_discovery
and this RFC:
http://tools.ietf.org/html/rfc6763
@philips @miekg

**Comment by geekgonecrazy** [comment]
@dapithor ah!  I think I get it now.  You're thinking more along the line of discovery as in discovering a server that is hosting the files.

From my understanding the objective is to operate like golang package discovery works.  It doesn't need to discover the server. Because the server is part of the path you enter.  So you'd enter something like `rkt fetch example.com/mycontainer` It then tries to determine where the aci file is located using meta tags if they are available or as this proposes using a dns record if available.

As far as it discovering a registry as it seems you're proposing.. That might be a completely different matter all together.

**Comment by dapithor** [comment]
I was thinking dual fold purpose allowing you to fetch an ACI without knowing storage location but the app URI:// name because if you know the URI and basic template default of {name}-{version}-{os}-{arch}.aci and that fails the SPEC says to move onto meta discovery (using logic to parse meta-tags).  My thoughts were that you just know you need [reduce-worker] and have no clue if it's on https, hdfs, sftp,etc but you can not only find out who out there is serving it and on what port and protocol but then correlate that with meta tags.  So, instead of html it would be in a TXT record.

` rkt fetch example.com/reduce-worker `
1. check https://{name}-{version}-{os}-{arch}.aci
2. Success GOTO fetch sig ELSE start ac meta discovery [line 3]
3. DNS request a SRV record for ACI service (get back a protocol, port, location)
4. DNS request a TXT for key value pair's
5. pull down ACI with your ACE app based on SRV and TXT data

DNS SRV:
> _ac-discovery._hdfs.example.com 86400 IN SRV 0 5 50070 dfs.example.com.

DNS TXT:
> name="ac-discovery" content="example.com {os}/{arch}/{name}-{version}.{ext}"
> name="ac-discovery" content="example.com {name}-{version}-{os}-{arch}.{ext}"

### Output producing:
> hdfs://dfs.example.com:50070/{os}/{arch}/{name}-{version}.{ext}

I understand this might be forward thinking or not needed (adding in a SRV) but my thought was having to always modify an html page of meta tags can be cumbersome versus if you move from foo.com to bar.com or change a port you don't have to go and change those html files independently or use a configuration management template to do so.  Sorry if I made this more complicated than needed to be!

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [spec] appc: use TXT dns record..." ]
>Fair points but application logic needs to happen regardless if one record or N records.  So, if you look up anything logic needs to be in the app but SRV/TXT etc being built into most DNS servers/services alleviates the need to also have a custom dns server for a custom RR.
>Even with zeroconf/dns-sd (which is basically what I am advocating and should have clarified)  you still need a logic/library and tell it to use it? Am I too far off base ?
>My thoughts were basically DNS-SD which is SRV/TXT in this wiki:
>http://en.wikipedia.org/wiki/Zero-configuration_networking#DNS-based_service_discovery
>and this RFC:
>http://tools.ietf.org/html/rfc6763
>@philips @miekg

DNS-SD needs client support as well, but maybe more widely deployed.

**Comment by rsertelon** [comment]
@dapithor I like your idea of using SRV + TXT.

A few things come to my mind:

1/ I would suggest that TXT records should be optional, in the sense that if none is specified a default lookup format should be used (as it is done for CardDAV/CalDAV). I believe that it is `{name}-{version}-{os}-{arch}.aci` that is specified for now.

2/ What if the procotol is `_tcp`? Should HTTP(s) be used by default? Should the client use HTTP(s) based on the specified port?

**Comment by b** [comment]
1) Storing this sort of data in TXT records happens frequently. For example, that's how SPF works. While a single TXT record is limited to 255 bytes of data, you can have multiple TXT records for a given label which can then be stitched together by clients. Keep in mind that any response over 512 bytes will cause DNS servers to truncate responses, making clients connect over TCP to ask the same query again, so best to keep things short.

2) Multiple lookups for service discovery is normal. MX, SRV, and CNAME records all require multiple resolutions to get from a name to an IP. The behavior is controlled by the clients, not the resolver. I would expect most of the lookups involved here would be in cache and this isn't an interactive scenario where a few extra milliseconds of DNS resolution is a concern.

3) The _proto field in an SRV record is exactly that: the protocol. In everything being discussed, that is TCP. Similarly, the _service field would be HTTP for what is being discussed. Distinguishing between HTTP and HTTPS would be down to the URI, not part of the SRV record. There is NO reason for a new RR type and going down that path would almost certainly stop anyone from deploying it.

4) DNS-SD, which actually uses 3 RR types, covers a lot of this, as has already been discussed. The exact data to be included in each record, the structure of names, the way names should be encoded and resolution performed needs to be defined as part of the app container spec since the DNS-SD specification leaves a lot of room for variation. I strongly support this standards-based approach.

**Comment by miekg** [comment]
Note that TXT records are *not* limited to 255 bytes, you can put 64K of data in a TXT record.
Also note that you can easily have messages larger than 512 bytes with EDNS0 bufsize, the *only* downside of large packets messages be UDP fragmentation

Usually a server is smart in that it will try to put the IP address for MX, SRV in the additional section of a message. In the case of CNAME, the target (if the servers has it) is put in the answer section.

There is all the reason for a new RR type, because that is the easy thing to do. Reusing TXT is just about the dumbest thing to do here. *BUT* getting a supported everything is a pain. I agree with that.

**Comment by b** [comment]
I understand how quickly these sorts of discussions devolve into protocol pedantry, but here we go. A given TXT RR in a zone file is limited to 255 bytes. A label may have multiple TXT RRs, totaling up to 64K. EDNS0 is not universally supported, so 512 bytes is the conservative choice certain to work as expected in all situations. For example, there are still many firewalls deployed that happily discard UDP DNS responses larger than 512 bytes.

While server might add data in the additional section, that is an optimization the number of round trips. It is still on the client, not the resolver, to pick out the records it needs and what to do with them. There is no reason servers couldn't or wouldn't piggyback the TXT records being discussed, just as they do with SPF and with things I, and no doubt many others, have implemented. It's not "dumb", it's using the tools as intended.

Asserting that there is a reason is not the same as demonstrating there is a reason. There is no suggestion the desired behavior cannot be achieved with existing RR types, only that some slight optimization in resolution might be possible at the cost of incompatibility with all deployed DNS implementations. SPF and DNS-SD are just two of many examples of how to properly use existing resource records in combination to achieve specific results and there is no reason this situation is any different.

**Comment by miekg** [comment]
I'm sorry, but regarding the length of TXT records you have it wrong:
https://www.ietf.org/rfc/rfc1035.txt, section "3.3.14. TXT RDATA format"

> TXT-DATA        One or more `<character-string>`s.

Where `<character-string>`s. is defined as:

> <character-string> is a single
> length octet followed by that number of characters.  <character-string>
> is treated as binary information, and can be up to 256 characters in
> length (including the length octet).

And yes firewalls would drop larger packets, with the advent of DNSSEC this has become less and less (luckily).

And yes, existing types would work, but in the longer run a new RR type would work just so much better. It is easy to do and any nameserver support unknown RRs (<https://tools.ietf.org/html/rfc3597>) can support it.

**Comment by b** [comment]
No, I have it right. That 256 characters including the length octet? That's 255 character strings, just like I said. TXT-DATA is one OR MORE of those 255 character strings. The limit on the length of RDATA is 65535. The RDATA for a TXT RR is up to 65535 octets of 255 octet strings, just like I said. I did not use the pedantic terminology because it is unnecessary to this discussion.

There are still an unfortunately large number of those older devices dropping EDNS0 responses. The default image discovery mechanism should not rely on an extension we know to not yet be universal. Suggesting we layer on an additional requirement for both EDNS0 _and_ unknown RRs is just asking for trouble.

Please quantify "would work so much better" given there are multiple IETF protocols that behave in exactly the way being proposed and which don't rely on anything but basic resolver behavior. If this is merely a question of personal aesthetics, then we can agree to disagree.

**Comment by geekgonecrazy** [comment]
Every dns provider i've tried limits it to 255 characters.. We can't just assume that because it **can technically** be done that it should.  

If i'm not mistaken compatibility is the key here.

**Comment by b** [comment]
Limits a single TXT record in the zone file to 255 characters, right?

**Comment by geekgonecrazy** [comment]
>> Limits a single TXT record in the zone file to 255 characters, right?

Correct.  I intentionally have avoided trying custom dns services run locally.  Because not everyone is going to be running their own dns.

**Comment by b** [comment]
You may be misunderstanding. That is the single 255 byte string per entry
part of what I said above. You can have multiple TXT records for the same
label. That is actually required by SPF.

**Comment by geekgonecrazy** [comment]
I articulated it poorly. But yes that is what I meant.  They allow for multiple 255 byte strings, or multiple TXT records.

Either way.  If its multiple chunks in one TXT record or multiple TXT records.  Both are easier for then a new RR.  Simply because they already exist and are widely used for this exact sort of thing.

If splitting it up into multiple records makes it easier for the DNS client to load.  I say this should be the recommended approach.

**Comment by geekgonecrazy** [comment]
Any more thoughts on doing it using TXT?  This has been stale for a while.  I'd love to see this implemented.

**Comment by jonboulle** [comment]
This is still on the cards as something we might consider implementing, but
work on discovery is a little stalled right now as we figure out the next
steps for appc + cncf.

**Comment by geekgonecrazy** [comment]
@jonboulle perfect.  I'll stay tuned.  Thanks for the update!
